### PR TITLE
tests: serialize pipeline file

### PR DESCRIPTION
**Pull request by skshetry**
This PR adds sorting for params keys as well. Before, it was only getting sorted for lockfile.
Rest are tests and a bit of refactoring that does not change anything.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here. If the CLI API is changed, I have updated [tab completion scripts](https://github.com/iterative/dvc/tree/master/scripts/completion).

* [x] ❌ I will check DeepSource, CodeClimate, and other sanity checks below. (We consider them recommendatory and don't expect everything to be addressed. Please fix things that actually improve code or fix bugs.)

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Part of #3693

### CODE REVIEW

File: dvc/stage/serialize.py [hunk 1]
```diff
@@ -70,16 +82,15 @@ def _serialize_params(params: List[ParamsDependency]):
         dump = param_dep.dumpd()
         path, params = dump[PARAM_PATH], dump[PARAM_PARAMS]
         if isinstance(params, dict):
-            k = list(params.keys())
+            k = sorted(params.keys())
```
Comment (skshetry) [inline dvc/stage/serialize.py:85]: We were not sorting params keys for pipeline file before, only for the lock. Now, the same behavior is for lockfile.

File: tests/unit/stage/test_serialize_pipeline_file.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        {"custom.yaml": ["baz", "pqrs", "wxyz"]},
+    ]
+
+
+@pytest.mark.parametrize(
+    "typ, extra",
+    [("plots", {"plot": True}), ("metrics", {"metric": True}), ("outs", {})],
+)
+def test_outs_and_outs_flags_are_sorted(dvc, typ, extra):
+    stage = create_stage(PipelineStage, dvc, deps=["input"], **kwargs)
```
Comment (skshetry) [inline tests/unit/stage/test_serialize_pipeline_file.py:108]: Maybe, we should implement `{metrics,plots}[_persist][_no_cache]` that are missing?

Comment (pared) [reply tests/unit/stage/test_serialize_pipeline_file.py:108]: @skshetry Could you elaborate?

Comment (skshetry) [reply tests/unit/stage/test_serialize_pipeline_file.py:108]: On `.dvc` and `dvc.yaml` files, we allow metrics and plots to be `cache: False` and `persist: True`. But, it cannot be set from outside (eg: `fill_stage_outputs` and `dvc run`).  So, I was just saying that maybe we should implement that for `run`.

Comment (efiop) [reply tests/unit/stage/test_serialize_pipeline_file.py:108]: @skshetry Let's not create those `dvc run` options for now, it is crowded there as it is. We clearly need to consider a different CLI approach, as `metrics_persist_no_cache` is just too ugly and doesn't scale if we introduce yet another option :slightly_smiling_face: We've been discussing something like using a comma to join those fstab-style, but, again, not worth bothering with that for now.

**Comment by efiop** [comment]
For the record: snap is failing for unrelated reasons.

**Pull request merged**